libs/@blockprotocol/type-system/rust/src/schema/data_type/constraint/mod.rs

```diff
@@ -61,7 +76,79 @@ pub enum ValueConstraints {
     AnyOf(AnyOfConstraints),
 }
 
-impl Constraint<JsonValue> for ValueConstraints {
+impl ValueConstraints {
+    /// Folds multiple constraints into fewer constraints.
+    ///
+    /// This function attempts to combine as many constraints as possible. If two constraints
+    /// cannot be fully merged, they are kept separate.
+    ///
+    /// The algorithm works as follows:
+    /// - It iterattes over all constraints
+    /// - for each constraint, it tries to merge them with the constraints that have already been
+    ///   merged from the previous iterations from left to right
+    /// - if a constraint cannot be fully merged, it is either combined with the next constraint or
+    ///   added to the list of constraints that have already been merged.
+    ///
+    /// # Errors
+    ///
+    /// If two constraints exclude each other, an error is returned.
+    pub fn fold_intersections(
+        schemas: impl IntoIterator<Item = Self>,
+    ) -> Result<Vec<Self>, Report<ResolveClosedDataTypeError>> {
+        schemas
+            .into_iter()
+            .map(Some)
+            .try_fold(Vec::<Self>::new(), |mut folded, mut constraints| {
+                folded = folded
```

Comment:
non-blocking: I wonder if it would be more readable using a loop instead of iterator combinators.

Reply:
I'd keep it as it is for now, but I hear you and I may going to rewrite it as for-loop when I have to add more features in this logic anyway.